### PR TITLE
Any backbone for linear segmentation

### PR DESCRIPTION
**Pull request by liopeer**
## What has changed and why?
This is the beginning of a refactoring effort
- adds a new way for configs
- allows easy extension of linear segmentation for more backbones

Follow-ups will rename things away from DINOv2 and to generic names. After that I will move on to object detection, but don't want this to interfere with #772 etc. (ECViT implementations).

## How has it been tested?
- I'll add unit tests later during clean-up (there are anyway renamings still missing etc.)

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [x] Not needed (internal change)

## Did you update the documentation?

- [ ] Yes
- [x] Not needed (internal change without effects for user)

### CODE REVIEW

File: src/lightly_train/_models/dinov2_vit/dinov2_vit_package.py [hunk 1]
```diff
@@ -31,12 +31,12 @@
...
 from lightly_train._models.model_wrapper import ModelWrapper
-from lightly_train._models.package import Package
+from lightly_train._models.package import ModelNameParser, MultiScaleFeaturePackage
 
 logger = logging.getLogger(__name__)
 
 
-class DINOv2ViTPackage(Package):
+class DINOv2ViTPackage(MultiScaleFeaturePackage, ModelNameParser):
     name = "dinov2"
```
[the remark this answers is not in the document]
Comment (liopeer) [reply src/lightly_train/_models/dinov2_vit/dinov2_vit_package.py]: We use it mainly for type checking. It's nicer to be able to use `isinstance` checks and type hints instead of relying on brittle `if hasattr("parse_model_name"): ...`.

File: src/lightly_train/_models/dinov3/dinov3_package.py [hunk 2]
```diff
@@ -217,7 +217,7 @@ class _DINOv3ModelInfo(TypedDict):
 }
 
 
-class DINOv3Package(Package):
+class DINOv3Package(MultiScaleFeaturePackage, ModelNameParser):
     name = "dinov3"
```
[the remark this answers is not in the document]
Comment (liopeer) [reply src/lightly_train/_models/dinov3/dinov3_package.py]: Same here.

File: src/lightly_train/_models/dinov3/dinov3_vit.py [hunk 3]
```diff
@@ -24,13 +24,11 @@
...
     ForwardPoolOutput,
-    MultiScaleFeatureModelWrapper,
+    MultiScaleFeatureViT,
 )
 
 
-class DINOv3ViTModelWrapper(
-    Module, MultiScaleFeatureModelWrapper, ArchitectureInfoGettable
-):
+class DINOv3ViTModelWrapper(Module, MultiScaleFeatureViT, ArchitectureInfoGettable):
```
[the remark this answers is not in the document]
Comment (liopeer) [reply src/lightly_train/_models/dinov3/dinov3_vit.py:31]: Same here.

File: src/lightly_train/_models/package.py [hunk 4]
```diff
@@ -64,3 +68,21 @@ def get_model(
     def get_model_wrapper(cls, model: PackageModel) -> ModelWrapper:
         """Wrap the underlying model with the ModelWrapper."""
         ...
+
+
+class MultiScaleFeaturePackage(Package):
```
Comment (liopeer) [inline src/lightly_train/_models/package.py:73]: Note: Eventually all packages should implement this.

File: src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/config.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+import logging
+from typing import Any
+
+from pydantic import Field
+
+from lightly_train._configs.config import ConfigsNamespace, PydanticConfig
+from lightly_train._configs.model_registry import ModelRegistry
+
+logger = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (liopeer) [reply src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/config.py]: Done.

File: src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/task_model.py [hunk 6]
```diff
@@ -22,17 +22,19 @@
 
 from lightly_train._data import file_helpers
 from lightly_train._models import package_helpers
-from lightly_train._models.dinov2_vit.dinov2_vit_package import DINOV2_VIT_PACKAGE
-from lightly_train._models.dinov2_vit.dinov2_vit_src.models.vision_transformer import (
-    DinoVisionTransformer,
+from lightly_train._models.model_wrapper import ModelWrapper
+from lightly_train._models.package import MultiScaleFeaturePackage
+from lightly_train._task_models.dinov2_linear_semantic_segmentation.config import (
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/task_model.py:27]: I'll do a lot of renaming from `dinov2_linear_semantic_...` to just `semantic_` in a follow up.

File: src/lightly_train/_task_models/linear_semantic_segmentation/train_model.py [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+)
+from lightly_train._data.task_data_args import TaskDataArgs
+from lightly_train._metrics.semantic_segmentation.task_metric import (
+    SemanticSegmentationTaskMetric,
+    SemanticSegmentationTaskMetricArgs,
+)
+from lightly_train._optim import optimizer_helpers
+from lightly_train._task_models.linear_semantic_segmentation.task_model import (
+    LinearSemanticSegmentation as DINOv2LinearSemanticSegmentation,
+)
```
[the remark this answers is not in the document]
Comment (liopeer) [reply src/lightly_train/_task_models/linear_semantic_segmentation/train_model.py]: Will be addressed later.

File: src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/config.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+    @LINEAR_SEG_MODEL_REGISTRY.register(
+        "dinov2/vits14-linear",
+        "dinov2/vits14-notpretrained-linear",
+    )
+    class DINOv2ViTS14(LinearSemanticSegmentationConfig):
+        backbone_name: str = "dinov2/vits14"
+        backbone_args: dict[str, Any] = Field(
+            default_factory=lambda: {"drop_path_rate": 0.0}
+        )
+        freeze_mask_token: bool = True
```
Comment (mrpositron) [inline src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/config.py:51]: `dinov2/vits14-notpretrained-linear` option loads pretrained weights, because its config sets `backbone_name="dinov2/vits14"` (and it has a weights url) with `load_weights` defaulting to `True`. The fix will be to use `backbone_name="dinov2/vits14-notpretrained"` or a bit hacky to set `load_weights=False`.

File: src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/task_model.py [hunk 9]
```diff
@@ -144,52 +159,28 @@ def __init__(
...
+        # Fallback: accept structurally valid names with a supported MultiScaleFeaturePackage.
+        if not model.endswith(f"-{cls.model_suffix}"):
+            return False
+        backbone_part = model[: -len(f"-{cls.model_suffix}")]
         try:
-            backbone_name = DINOV2_VIT_PACKAGE.parse_model_name(
-                model_name=backbone_name
-            )
+            package_name, _ = package_helpers.parse_model_name(backbone_part)
+            package = package_helpers.get_package(package_name)
```
Comment (mrpositron) [inline src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/task_model.py:192]: Should we check if the model `model_name` is also supported? Otherwise it will return error later in `src/lightly_train/_models/dinov2_vit/dinov2_vit_package.py`, in `model_info = VIT_MODELS.get(model_name)`.

File: src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/config.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    @LINEAR_SEG_MODEL_REGISTRY.register("dinov3/convnext-small-eupe-linear")
+    class DINOv3ConvNextSmallEupe(LinearSemanticSegmentationConfig):
+        backbone_name: str = "dinov3/convnext-small-eupe"
+
+    @LINEAR_SEG_MODEL_REGISTRY.register("dinov3/convnext-base-eupe-linear")
+    class DINOv3ConvNextBaseEupe(LinearSemanticSegmentationConfig):
+        backbone_name: str = "dinov3/convnext-base-eupe"
+
+    class Fallback(LinearSemanticSegmentationConfig):
+        pass  # backbone_name stays "", parsed from model_name at runtime
```
Comment (mrpositron) [inline src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/config.py:208]: For valid but unregistered DINOv2 names like `dinov2/vits14-noreg-linear`, we hit `Fallback`, which leaves `freeze_mask_token=False`; previously we always froze it with `self.backbone.mask_token.requires_grad = False`. Since the mask token is unused in this forward path, this can reintroduce DDP unused-parameter issues during full fine-tuning.

See the comment there (src/lightly_train/_task_models/linear_semantic_segmentation/task_model.py, line 142)
```
# TODO(Guarin, 07/25): Improve how mask tokens are handled for fine-tuning.
# Should we drop them from the model? We disable grads here for DDP to work
# without find_unused_parameters=True.
```

File: src/lightly_train/_models/package.py (hunk 4 above)
Comment (liopeer) [reply src/lightly_train/_models/package.py:73]: Created issues.

File: src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/task_model.py (hunk 6 above)
Comment (liopeer) [reply src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/task_model.py:27]: Done in the follow-up PR.

File: src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/config.py (hunk 8 above)
Comment (liopeer) [reply src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/config.py:51]: Adressed it, they now have their own configs.

File: src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/config.py (hunk 10 above)
Comment (liopeer) [reply src/lightly_train/_task_models/dinov2_linear_semantic_segmentation/config.py:208]: Thx, I am actually adding a unit test that makes sure that all models are always registered.

File: tests/_task_models/dinov2_linear_semantic_segmentation/test_config.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    ]
+
+
+def _registered_config(model_name: str) -> type[LinearSemanticSegmentationConfig]:
+    return LINEAR_SEG_MODEL_REGISTRY.get(
+        model_name, default=LinearSegConfigRegistry.Fallback
+    )
+
+
+class TestLinearSegConfigRegistry:
```
Comment (liopeer) [inline tests/_task_models/dinov2_linear_semantic_segmentation/test_config.py]:
```suggestion
class TestLinearSegModelRegistry:
```

File: src/lightly_train/_task_models/linear_semantic_segmentation/task_model.py [hunk 12]
```diff
@@ -0,0 +1,493 @@
...
+                image_sizes=[(meta["resized_h"], meta["resized_w"])],
+            )[0]  # (K|K+1, H', W')
+
+            if self.class_ignore_index is not None:
+                # Restrict to known classes only.
+                logits = logits[:-1]
+            logits = logits.unsqueeze(0)  # (1, K, H', W')
+            logits = F.interpolate(
+                logits, size=(meta["orig_h"], meta["orig_w"]), mode="bilinear"
+            )
```
[the remark this answers is not in the document]
Comment (mrpositron) [reply src/lightly_train/_task_models/linear_semantic_segmentation/task_model.py]: I think this copilot comment is relevant

**Review by mrpositron — APPROVED** [review]
LGTM 🚀
